src/hybridmap/HybridNeutralAtomMapper.cpp

```diff
@@ -0,0 +1,1293 @@
+  fp distanceChangeLookahead = 0;
+  if (!this->lookaheadLayerGate.empty()) {
+    distanceChangeLookahead =
+        swapCostPerLayer(swap, swapCloseByLookahead, swapExactLookahead) /
+        static_cast<fp>(this->lookaheadLayerGate.size());
+  }
+  auto cost = parameters.lookaheadWeightSwaps * distanceChangeLookahead +
+              distanceChangeFront;
+  //  compute the last time one of the swap qubits was used
+  if (this->parameters.decay != 0) {
+    uint32_t idxLastUsed = 0;
+    for (uint32_t i = 0; i < this->lastBlockedQubits.size(); ++i) {
+      if (this->lastBlockedQubits[i].find(swap.first) !=
+              this->lastBlockedQubits[i].end() ||
+          this->lastBlockedQubits[i].find(swap.second) !=
+              this->lastBlockedQubits[i].end()) {
+        idxLastUsed = i;
+        break;
+      }
+    }
+    cost *= this->decayWeights[idxLastUsed];
+  }
+  return cost;
+}
+
+std::pair<Swaps, WeightedSwaps>
+NeutralAtomMapper::initSwaps(const GateList& layer) {
+  Swaps         swapCloseBy = {};
+  WeightedSwaps swapExact   = {};
+  // computes for each gate the necessary moves to execute it
+  for (const auto& gate : layer) {
+    auto usedQubits   = gate->getUsedQubits();
+    auto usedHwQubits = this->mapping.getHwQubits(usedQubits);
+    if (usedQubits.size() == 2) {
+      // swap close by for two qubit gates
+      swapCloseBy.emplace_back(*usedHwQubits.begin(), *usedHwQubits.rbegin());
+    } else {
+      // for multi-qubit gates, find the best position around the gate qubits
+      auto bestPos = getBestMultiQubitPosition(gate);
+      if (verbose) {
+        std::cout << "bestPos: ";
+        for (auto qubit : bestPos) {
+          std::cout << qubit << " ";
+        }
+        std::cout << '\n';
+      }
+      // then compute the exact moves to get to the best position
+      auto exactSwapsToPos = getExactSwapsToPosition(gate, bestPos);
+      swapExact.insert(swapExact.end(), exactSwapsToPos.begin(),
+                       exactSwapsToPos.end());
+    }
+  }
+  // sort and remove duplicates from moveExact
+  swapExact.erase(std::unique(swapExact.begin(), swapExact.end(),
+                              [](const auto& swap1, const auto& swap2) {
+                                return swap1.first == swap2.first;
+                              }),
+                  swapExact.end());
+  return {swapCloseBy, swapExact};
+}
+
+fp NeutralAtomMapper::swapCostPerLayer(const Swap&          swap,
+                                       const Swaps&         swapCloseBy,
+                                       const WeightedSwaps& swapExact) {
+  fp distBefore = 0;
+  fp distAfter  = 0;
+  fp distChange = 0;
+  // bring close only until swap distance =0, bring exact to the exact position
+  // bring qubits together to execute gate
+  for (const auto& [q1, q2] : swapCloseBy) {
+    // distance before
+    distBefore = this->hardwareQubits.getSwapDistance(q1, q2);
+    if (distBefore == std::numeric_limits<fp>::infinity()) {
```

Comment:
Stated this elsewhere already, but just to be sure: prefer to use finite values, e.g. `max()`. Please also check throughout this file